### PR TITLE
Fetching  & passing FormXml to improve copilot response accuracy

### PR DESCRIPTION
**Pull request by amitjoshi438**
**This pull request primarily focuses on fetching FORM_XML for copilot scenarios. The fetched FormXML is parsed to get info about the form and enhance the context passing to copilot.**

Dependency updates:
* [`package.json`](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R1025-R1026): Added `@microsoft/1ds-core-js` and `@microsoft/1ds-post-js` to `devDependencies`. Moved `@microsoft/generator-powerpages` and `@types/jwt-decode` from `dependencies` to `devDependencies`. Added `@types/xmldom` and `xmldom` to `dependencies`. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R1025-R1026) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L1083-R1094) [[3]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R1115)

Code refactoring:
* [`src/common/copilot/IntelligenceApiService.ts`](diffhunk://#diff-a61cdf8af8b4c524990ec9dfb07752b80f4e7dad3fae3a337383fe151d79c176R24): Added a line to disable the eslint rule `prefer-const` for the `requestBody` variable.
* [`src/common/copilot/PowerPagesCopilot.ts`](diffhunk://#diff-98c063ddf20a69c9594332e9f968fe00b2e009c5de85256f9e336c089e6880abL23-R23): Imported `getFormXml` from `dataverseMetadata`. Refactored the `PowerPagesCopilot` class to use the new `getFormXml` function and to handle changes in the return type of `getEntityName`. [[1]](diffhunk://#diff-98c063ddf20a69c9594332e9f968fe00b2e009c5de85256f9e336c089e6880abL23-R23) [[2]](diffhunk://#diff-98c063ddf20a69c9594332e9f968fe00b2e009c5de85256f9e336c089e6880abL335-R352)

New functions:
* [`src/common/copilot/dataverseMetadata.ts`](diffhunk://#diff-f01824da34af02017a75c331f96e8098e03b6d9c20d04bf943d8551fddde621fR14): Imported `DOMParser` from `xmldom`. Added new functions `getFormXml` and `parseXML` to parse XML data from form responses. Updated the `getEntityName` function to return an object with `entityName` and `formName` properties. [[1]](diffhunk://#diff-f01824da34af02017a75c331f96e8098e03b6d9c20d04bf943d8551fddde621fR14) [[2]](diffhunk://#diff-f01824da34af02017a75c331f96e8098e03b6d9c20d04bf943d8551fddde621fR48-R73) [[3]](diffhunk://#diff-f01824da34af02017a75c331f96e8098e03b6d9c20d04bf943d8551fddde621fR94-R138) [[4]](diffhunk://#diff-f01824da34af02017a75c331f96e8098e03b6d9c20d04bf943d8551fddde621fR158-R169)

Minor changes:
* [`src/common/copilot/welcome-notification/CopilotNotificationPanel.ts`](diffhunk://#diff-65e49faa3d1152985263206facedc615bea8a1e492032f141378741f938f8cf2L23-L24): Removed a console log statement.

### CODE REVIEW

File: src/common/copilot/IntelligenceApiService.ts [hunk 1]
```diff
@@ -15,12 +15,13 @@ import { EXTENSION_NAME } from "../../client/constants";
 const clientType = EXTENSION_NAME + '-' + getExtensionType();
 const clientVersion = getExtensionVersion();
 
-export async function sendApiRequest(userPrompt: UserPrompt[], activeFileParams: IActiveFileParams, orgID: string, apiToken: string, sessionID: string, entityName: string, entityColumns: string[], telemetry: ITelemetry, aibEndpoint: string | null, geoName: string | null) {
+export async function sendApiRequest(userPrompt: UserPrompt[], activeFileParams: IActiveFileParams, orgID: string, apiToken: string, sessionID: string, entityName: string, entityColumns: string [], telemetry: ITelemetry, aibEndpoint: string | null, geoName: string | null) {
```
Comment (ritikramuka) [inline src/common/copilot/IntelligenceApiService.ts:18]: NIT: Can you remove this added whitespace in order to consistency throughout the code.

File: src/common/copilot/PowerPagesCopilot.ts [hunk 2]
```diff
@@ -335,17 +335,24 @@ export class PowerPagesCopilot implements vscode.WebviewViewProvider {
 
                 this.sendMessageToWebview({ type: 'userName', value: userName });
 
-                let entityName = "";
-                let entityColumns: string[] = [];
+                let metadataInfo = { entityName: '', formName: '' };
+                let entityInfo : string [] = [];
```
Comment (ritikramuka) [inline src/common/copilot/PowerPagesCopilot.ts:339]: NIT: Can you remove this added whitespace in order to consistency throughout the code.

Comment (ritikramuka) [inline src/common/copilot/PowerPagesCopilot.ts:339]: Can a better name be added here? as entityInfo is also an interface having entityId, entityName and rootPageId

File: src/common/copilot/dataverseMetadata.ts [hunk 3]
```diff
@@ -44,6 +45,32 @@ export async function getEntityColumns(entityName: string, orgUrl: string, apiTo
     }
 }
 
+export async function getFormXml(entityName: string, formName: string,  orgUrl: string, apiToken: string, telemetry: ITelemetry, sessionID: string): Promise<(string [])>{
+    try {
+        const dataverseURL = `${orgUrl.endsWith('/') ? orgUrl : orgUrl.concat('/')}api/data/v9.2/systemforms?$filter=objecttypecode eq '${entityName}' and name eq '${formName}'`;
```
Comment (ritikramuka) [inline src/common/copilot/dataverseMetadata.ts:50]: let's move this to constants. and can we use `URL method` to create URL.

File: src/common/copilot/dataverseMetadata.ts [hunk 4]
```diff
@@ -64,9 +91,51 @@ function getAttributesFromResponse(jsonResponse: any): Attribute[] {
...
+    const parser = new DOMParser();
+    const xmlDoc = parser.parseFromString(formXml, "text/xml");
+
+    // Get all 'row' elements
+    const rows = xmlDoc.getElementsByTagName('row');
+
+    const result = [];
+
+    // Iterate over all 'row' elements
+    for (let i = 0; i < rows.length; i++) {
```
Comment (ritikramuka) [inline src/common/copilot/dataverseMetadata.ts:113]: NIT: can we use forEach loop?

File: src/common/copilot/dataverseMetadata.ts [hunk 5]
```diff
@@ -64,9 +91,51 @@ function getAttributesFromResponse(jsonResponse: any): Attribute[] {
...
+        const row = rows[i];
+
+        // Get the 'label' and 'control' elements within the current 'row'
+        const label = row.getElementsByTagName('label')[0];
+        const control = row.getElementsByTagName('control')[0];
+
+        // If both 'label' and 'control' elements exist, create an object and add it to the result array
+        if (label && control) {
+            const description = label.getAttribute('description');
+            const datafieldname = control.getAttribute('datafieldname');
```
Comment (ritikramuka) [inline src/common/copilot/dataverseMetadata.ts:123]: and if possible, can this be also added in constants

**Review by ritikramuka — CHANGES_REQUESTED** [review]

File: src/common/copilot/dataverseMetadata.ts (hunk 4 above)
Comment (amitjoshi438) [reply src/common/copilot/dataverseMetadata.ts:113]: 'forEach' does not exist on type 'HTMLCollectionOf<Element>'

Comment (ritikramuka) [reply src/common/copilot/dataverseMetadata.ts:120]: You can use Array.from and then forEach

File: src/common/copilot/dataverseMetadata.ts (hunk 3 above)
Comment (amitjoshi438) [reply src/common/copilot/dataverseMetadata.ts]: Whole URL can't be moved to constants as it contains dynamic values. Moved the` apiPath` to constants.

**Review by gshivi — APPROVED** [review]

**Review by BidishaMS — APPROVED** [review]

**Review by ritikramuka — APPROVED** [review]

**Pull request merged**